### PR TITLE
Adding method to allow search for Id

### PR DESCRIPTION
**Pull request by jayudey-wf**
Internal Review
## Issue
- There is no way to search for elements by Id
## Changes

**Source:**
- Added method to allow for this type of search

**Tests:**
- Added corresponding test
## Areas of Regression
- n/a
## Testing
- Ensure that tests pass
## Code Review

@greglittlefield-wf
@tylersnavely-wf
@jonasray-wf

### CODE REVIEW

File: test/react_test_utils_test.dart [hunk 1]
```diff
@@ -229,6 +229,22 @@ void main() {
     expect(results[1]['tagName'], equals('DIV'));
   });
 
+  test('scryRenderedDomComponentsWithId', () {
+    var idName = 'searchableId';
+
+    component = renderIntoDocument(div({}, [
+      div({'id': idName}),
+      div({'id': idName}),
+      span({})
```
Comment (tylersnavely-wf) [inline test/react_test_utils_test.dart]: It wouldn't be a bad idea to add another element with a different ID to verify that not all elements with IDs are being returned.

File: test/react_test_utils_test.dart [hunk 2]
```diff
@@ -229,6 +229,22 @@ void main() {
...
+    ]));
+
+    var results = scryRenderedDOMComponentsWithId(component, idName);
+
+    expect(results.length, 2);
+    expect(results[0]['props']['id'], equals(idName));
+    expect(results[1]['props']['id'], equals(idName));
+  });
```
Comment (tylersnavely-wf) [inline test/react_test_utils_test.dart:262]: It would also be a good idea to test the scenario where the query returns an empty list.

File: lib/react_test_utils.dart [hunk 3]
```diff
@@ -344,6 +344,20 @@ List scryRenderedDOMComponentsWithClass(JsObject tree, String className) {
 }
 
 /// Finds all instances of components in the rendered tree that are DOM
+/// components with an id matching id arguments value.
+List scryRenderedDOMComponentsWithId(JsObject tree, String id) {
+  dartTestFunction(window, component) {
+    if (component['props']['id'] == id) {
+      return true;
+    }
```
Comment (tylersnavely-wf) [inline lib/react_test_utils.dart]: I haven't tried, but can this be shortened to just one line?

```
dartTestFunction(window, component) => component['props']['id'] == id;
```

**Comment by greglittlefield-wf** [comment]
Why is this method needed? 

If it's to fetch specific rendered components for testing, it could be problematic when dealing with composite components that forward the `'id'` prop, to be rendered by DOM components:

For example, consider the following composite component usage (using JSX for simplicity):

``` jsx
<CompositeComponent id="myComponent">content</CompositeComponent>
```

This renders the following DOM:

``` jsx
<div id="myComponent">content</div>
```

via the following React element hierarchy:

``` jsx
<CompositeComponent> // props: {'id': 'myComponent'}
  <div> // props: {'id': 'myComponent'}
    content
  </div>
</CompositeComponent>
```

If you called `scryRenderedDomComponentsWithId`, it would return both the `CompositeComponent` instance and the `div` instance.

---

As a side note, to fetch specific instances of rendered components in Web Skin Dart tests, we use `refs` via a custom [`getRef(String ref)`](https://github.com/Workiva/web_skin_dart/blob/0.0.3/test/test_util/react_util.dart#L36-L42) function.

Here's an example in our tests: https://github.com/Workiva/web_skin_dart/tree/0.0.3/test/ui_components/progress/progress_bar_test.dart#L73

**Comment by jayudey-wf** [comment]
i thought it was a helpful method because generally during functional testing id's are utilized to identify elements that the webdriver is trying to interact with but since there is no method to search for id's in the test utils there are times where id's and classnames' are added just for testing purposes to facilitate functional testing and unit testing with react test utils

i wrote this while working with a team to implement some react testing and thought it would be helpful to share

**Comment by greglittlefield-wf** [comment]
Wouldn't a use-case like that be more suited to using something like `Element.querySelector`?

**Comment by jayudey-wf** [comment]
closing pr as one of the main use cases is resolved by https://github.com/cleandart/react-dart/pull/66/
